code/modules/xenoarchaeology/xenoartifact_labeler.dm

```diff
@@ -0,0 +1,247 @@
+/obj/item/xenoartifact_labeler
+	name = "artifact labeler"
+	icon = 'icons/obj/xenoarchaeology/xenoartifact_tech.dmi'
+	icon_state = "xenoartifact_labeler"
+	desc = "A tool scientists use to label their alien bombs."
+	throw_speed = 3
+	throw_range = 5
+	w_class = WEIGHT_CLASS_TINY
+
+	var/list/activator = list() //Checked trait
+	var/list/activator_traits = list() //Display names
+
+	var/list/minor_trait = list()
+	var/list/minor_traits = list()
+
+	var/list/major_trait = list()
+	var/list/major_traits = list()
+
+	var/list/malfunction = list()
+	var/list/malfunction_list = list()  
+
+	var/list/info_list = list() //trait dialogue essentially
+
+	var/sticker_name //Name artifacts something pretty
+	var/list/sticker_traits = list()//passed down to sticker
+
+/obj/item/xenoartifact_labeler/Initialize()
+	. = ..()
+
+/obj/item/xenoartifact_labeler/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "XenoartifactLabeler")
+		ui.open()
+
+/obj/item/xenoartifact_labeler/ui_data(mob/user)
+	var/list/data = list()
+	data["activator"] = activator
+	data["activator_traits"] = get_trait_list_desc(activator_traits, /datum/xenoartifact_trait/activator)
+
+	data["minor_trait"] = minor_trait
+	data["minor_traits"] = get_trait_list_desc(minor_traits, /datum/xenoartifact_trait/minor)
+
+	data["major_trait"] = major_trait
+	data["major_traits"] = get_trait_list_desc(major_traits, /datum/xenoartifact_trait/major)
+
+	data["malfunction"] = malfunction
+	data["malfunction_list"] = get_trait_list_desc(malfunction_list, /datum/xenoartifact_trait/malfunction)
+
+	data["info_list"] = info_list
+
+	return data
+
+/obj/item/xenoartifact_labeler/ui_act(action, params)
+	if(..())
+		return
+	//Switch function could work here but it'd be pointless for the 1-2 cases
+	if(action == "print_traits")
+		create_label(sticker_name)
+		return
+
+	if(action == "clear_traits")
+		clear_selection()
+		return
+
+	if(action == "change_print_name" && istext(params["name"]))
+		sticker_name = sanitize_text("[params["name"]]")
+		return
+
+	trait_toggle(action, "activator", activator_traits, activator)
+	trait_toggle(action, "minor", minor_traits, minor_trait)
+	trait_toggle(action, "major", major_traits, major_trait)
+	trait_toggle(action, "malfunction", malfunction_list, malfunction)
+
+	. = TRUE
+	update_icon()
+
+/obj/item/xenoartifact_labeler/proc/get_trait_list_desc(list/traits, trait_type)//Get a list of all the specified trait types names, actually
+	trait_type = typesof(trait_type)
+	for(var/t in trait_type)
+		var/datum/xenoartifact_trait/X = t
+		if(initial(X.desc) && !(initial(X.desc) in traits) && !(initial(X.label_name)))
+			traits += initial(X.desc)
+		else if(initial(X.label_name) && !(initial(X.label_name) in traits)) //For cases where the trait doesn't have a desc or is tool cool to use one
+			traits += initial(X.label_name)
+	return traits
```

Comment:
Where does this traits var come from? Lists are references, so you are adding to whatever list is passed in here

Reply:
Exactly, you assign a list you want to feed it into.
``get_trait_list_desc(malfunction_list, /datum/xenoartifact_trait/malfunction)``
I will change it since it seems janky.